components/region_cache_memory_engine/src/engine.rs

```diff
@@ -73,22 +152,46 @@ pub struct RegionCacheMemoryEngine {
 }
 
 impl Debug for RegionCacheMemoryEngine {
-    fn fmt(&self, _: &mut fmt::Formatter<'_>) -> fmt::Result {
-        unimplemented!()
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "Region Cache Memory Engine")
     }
 }
 
 impl RegionCacheMemoryEngine {
-    pub fn new() -> Self {
-        RegionCacheMemoryEngine::default()
+    pub fn new_region(&self, region_id: u64) {
+        let mut core = self.core.lock().unwrap();
+
+        assert!(core.engine.get(&region_id).is_none());
+        assert!(core.region_metas.get(&region_id).is_none());
+        core.engine.insert(region_id, RegionMemoryEngine::default());
+        core.region_metas
+            .insert(region_id, RegionMemoryMeta::default());
     }
 }
 
 impl RegionCacheEngine for RegionCacheMemoryEngine {
     type Snapshot = RegionCacheSnapshot;
 
     fn snapshot(&self, region_id: u64, read_ts: u64) -> Option<Self::Snapshot> {
-        unimplemented!()
+        let mut core = self.core.lock().unwrap();
+        let region_meta = core.region_metas.get_mut(&region_id)?;
+        if !region_meta.can_read {
+            return None;
+        }
+
+        if read_ts <= region_meta.safe_ts {
+            // todo(SpadeA): add metrics for it
+            return None;
+        }
+
+        region_meta.snapshot_list.new_snapshot(read_ts);
+
+        Some(RegionCacheSnapshot {
```

Comment:
it's better to have a RegionCacheSnapshot::new and then put  region_meta.snapshot_list.new_snapshot(read_ts) inside the new(). So both new_snapshot and reclaim_snapshot are called in the same place.

Reply:
I don't understand. How can we "put region_meta.snapshot_list.new_snapshot(read_ts) inside the new()"?